src/connectors/queue/revision.ts

```diff
@@ -0,0 +1,310 @@
+
+      // Step 1: checks
+      if (!draft) {
+        job.progress(100)
+        done(null, `Revision draft ${draftId} not found`)
+        return
+      }
+      if (draft.publishState !== PUBLISH_STATE.pending) {
+        job.progress(100)
+        done(null, `Publication of revision draft ${draftId} is not pending.`)
+        return
+      }
+      const article = await this.articleService.baseFindById(draft.articleId)
+      if (!article) {
+        job.progress(100)
+        done(null, `Revised article ${draft.articleId} not found`)
+        return
+      }
+      if (article.state !== ARTICLE_STATE.active) {
+        job.progress(100)
+        done(null, `Revised article ${draft.articleId} is not active`)
+        return
+      }
+      const preDraft = await this.draftService.baseFindById(article.draftId)
+      job.progress(10)
+
+      // Step 2: publish content to IPFS, and handles error including notification
+      let ipfs: any
+      try {
+        ipfs = await this.articleService.publishToIPFS({
+          ...draft,
+          summary: draft.summary || makeSummary(draft.content),
+        })
+
+        if (!ipfs.dataHash || !ipfs.mediaHash) {
+          throw new Error(
+            `failed to publish content of draft ${draftId} to IPFS`
+          )
+        }
+      } catch (error) {
+        await this.draftService.baseUpdate(draft.id, {
+          publishState: PUBLISH_STATE.error,
+        })
+
+        this.notificationService.trigger({
+          event: 'revised_article_not_published',
+          recipientId: article.authorId,
+          entities: [
+            {
+              type: 'target',
+              entityTable: 'article',
+              entity: article,
+            },
+          ],
+        })
+        throw error
+      }
+      job.progress(30)
+
+      // Step 3: mark draft as published and copy data back to article
+      await this.draftService.baseUpdate(draft.id, {
+        dataHash: ipfs.dataHash,
+        mediaHash: ipfs.mediaHash,
+        archived: true,
+        publishState: PUBLISH_STATE.published,
+        updatedAt: new Date(),
+      })
+
+      const updatedArticle = await this.articleService.baseUpdate(article.id, {
+        dataHash: ipfs.dataHash,
+        mediaHash: ipfs.mediaHash,
+        draftId: draft.id,
+      })
+      job.progress(50)
+
+      // Step 4: transfer previous draft assets to current draft
+      // Note: collection and tags are handled in edit resolver.
+      // @see src/mutations/article/editArticle.ts
+      const {
+        id: entityTypeId,
+      } = await this.systemService.baseFindEntityTypeId('draft')
+      const assetMapIds = (
+        await this.systemService.findAssetAndAssetMap({
+          entityTypeId,
+          entityId: preDraft.id,
+        })
+      ).map((assetMap) => assetMap.id)
+      await this.systemService.swapAssetMapEntity(
+        assetMapIds,
+        entityTypeId,
+        draft.id
+      )
+      job.progress(60)
```

Comment:
@guoliu I'm not sure we should transfer previous draft's assets to new one. The only advantage is if we allow user to drop previous drafts, then assets can be kept. Any opinion?

cc @robertu7

Reply:
@devformatters I think we shouldn't allow newer draft (or article version) to inherit assets from older ones, as long as the newer draft already have the assets it needs. So that when we do need to delete drafts, we can cascade delete assets that are not used by other drafts.

So I'm not sure about this part here: if in other places we already assigned the needed assets to the new draft, we don't need this part, but if we don't then we still need to keep it.

Reply:
Based on last discussion, current revision feature won't allow users to change embed assets. So, the simplest way is to copy asset maps. I've changed codes.